### PR TITLE
Additional Configurability of FlatFileStorage

### PR DESCRIPTION
**Pull request by Wealthyturtle**
### Description
Allows users to configure the variable change threshold to trigger a re-save of the database file, as well as how often the database file is re-written.

---
**Target Minecraft Versions:**     Any
**Requirements:**     None
**Related Issues:**     #2007

### CODE REVIEW

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java [hunk 1]
```diff
@@ -191,9 +194,38 @@ protected boolean load_i(final SectionNode n) {
 			Skript.info(file.getName() + " successfully updated.");
 		}
 		
+		String thresholdRaw = getValue(n, "variable re-save threshold");
```
Comment (bensku) [inline src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Use getValue overload that takes in type.

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java [hunk 2]
```diff
@@ -191,9 +194,38 @@ protected boolean load_i(final SectionNode n) {
...
+				if (threshold < 0)
+					Skript.error("The variable re-save threshold cannot be negative!");
+				else
+					REQUIRED_CHANGES_FOR_RESAVE = threshold;
+			} catch (NumberFormatException e) {
+				Skript.error("Invalid integer '" + thresholdRaw);
+			}
+		}
+		
+		String rewriteRaw = getValue(n, "file re-write frequency in ticks");
```
Comment (bensku) [inline src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Same here.

File: src/main/resources/config.sk [hunk 3]
```diff
@@ -265,6 +265,15 @@ databases:
...
 		# type: SQLite
+		
+		variable re-save threshold: 1000
+		# The number of variables that are required to change in order for a re-save to occur.
+		# Setting this value lower is not recommended unless the server stores very few Skript variables, in which case it may be wise to lower
+		# this threshold to ensure that variables are always saved (prevents data loss during a crash).
+		
+		file re-write frequency in ticks: 6000
+		# The number of ticks that should pass before the database file is re-written.
+		# Setting this value lower is not recommended as saving large database files may cause server lag.
```
Comment (bensku) [inline src/main/resources/config.sk]: Comment should explicitly tell how low it is safe to set. Also, I don't think the description about lag is quite right. Saving large databases takes *time*, and bad things happen if a database save has not succeeded before next one starts.

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java [hunk 4]
```diff
@@ -99,6 +105,7 @@ protected boolean load_i(final SectionNode n) {
 		boolean update2_0_beta3 = false;
 		final Version v2_1 = new Version(2, 1);
 		boolean update2_1 = false;
+		final Version v_2_4_beta9 = new Version(2, 4, "beta 9");
```
Comment (bensku) [inline src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: This PR is staged for Skript 2.5. Just check for 2.4 or older version.

File: src/main/resources/config.sk (hunk 3 above)
Comment (Blueyescat) [reply src/main/resources/config.sk]: >  bad things happen if a database save has not succeeded before next one starts.

that should be prevented

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java [hunk 5]
```diff
@@ -84,7 +90,7 @@ protected FlatFileStorage(final String name) {
 	/**
 	 * Doesn'ts lock the connection as required by {@link Variables#variableLoaded(String, Object, VariablesStorage)}.
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Doesn't

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java [hunk 6]
```diff
@@ -191,9 +198,62 @@ protected boolean load_i(final SectionNode n) {
...
+			try {
+				Skript.info("Updating your config to version 2.4-beta9...");
+				r = new BufferedReader(new InputStreamReader(new FileInputStream(SkriptConfig.getConfig().getFile()), UTF_8));
+				List<String> convertedCfg = new ArrayList<>();
+				String line = null;
+				while((line = r.readLine()) != null) {
+					convertedCfg.add(line);
+					if (line.contains("type: CSV") || line.contains("type: SQLite")) {
+						convertedCfg.add("");
+						convertedCfg.add("		variable re-save threshold: 1000");
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: This option should come after the other one, :)

**Comment by Blueyescat** [comment]
I think the option names aren't user friendly.
```diff
- variable re-save threshold
+ required variable changes for save
```

and this should work with timespans
```diff
- file re-write frequency in ticks
+ variable save interval
```

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java [hunk 7]
```diff
@@ -191,9 +198,62 @@ protected boolean load_i(final SectionNode n) {
...
 		
+		if (SkriptConfig.getPreviousConfigVersion().isSmallerThan(v_2_4_beta9)) {
+			try {
+				Skript.info("Updating your config to version 2.4-beta9...");
+				r = new BufferedReader(new InputStreamReader(new FileInputStream(SkriptConfig.getConfig().getFile()), UTF_8));
+				List<String> convertedCfg = new ArrayList<>();
+				String line = null;
+				while((line = r.readLine()) != null) {
+					convertedCfg.add(line);
+					if (line.contains("type: CSV") || line.contains("type: SQLite")) {
```
Comment (Blueyescat) [inline src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Bad check

Comment (Blueyescat) [reply src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Wait, you are adding these options into the default storage setting section, but the saving code is global.
Don't forget that there can be multiple flat file storages (default is one of them with name 'default'), and a FlatFileStorage is equal to a single storage.

I mean, the saving code should be specific to a single storage (see how other storage specific options work), or at least the options should be in the global section.

Comment (Wealthyturtle) [reply src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Does the current code not work with that? Isn't a SectionNode referring to a section of the config and not the whole thing?

**Comment by Wealthyturtle** [comment]
Current Changes:
- Updated documentation of SkriptConfig.getPreviousConfigVersion()
- Added nullness to SkriptConfig.getPreviousConfigVersion(), in the case that for whatever reason the previous config version could not be derived
- Fixed typo in documentation of FlatFileStorage.load_i()
- Replaced null suppression with resource suppression in FlatFileStorage.load_i()
- Added resource suppression in FlatFileStorage.saveVariables() to make the file warning-less
- Updated previous config version testing to before version 2.5
- Renamed new options 'variable re-save threshold' and 'file re-write frequency in ticks' to 'required variable changes for save' and 'variable save interval' respectively
- Rewrote description of the 'required variable changes for save' and 'variable save interval' options to mention flat file databases
- Added a disclaimer for the 'variable save interval' option that setting a value to 1200 ticks or lower is not advisable as it may cause bad things to happen if a database save has not succeeded before next one starts

**Comment by bluelhf** [comment]
@Wealthyturtle Any updates on this?

File: src/main/java/ch/njol/skript/variables/FlatFileStorage.java (hunk 7 above)
Comment (Wealthyturtle) [reply src/main/java/ch/njol/skript/variables/FlatFileStorage.java]: Just tested, this code supports multiple flat file databases, with each database having its own save interval and min. variable threshold. However, I do agree that the check has issues and needs to be reworked.

**Comment by FranKusmiruk** [comment]
damn, tried to resolve conflicts and somehow this happened, going to try to fix it

**Comment by FranKusmiruk** [comment]
Closing in favour of #3786

**Pull request closed without merging**